numpyro/contrib/funsor/infer_util.py

```diff
@@ -70,6 +71,33 @@ def config_fn(site):
     return infer_config(fn, config_fn)
 
 
+def compute_markov_factors(time_to_factors, time_to_init_vars, time_to_markov_dims, sum_vars, prod_vars):
+    """
+    :param dict time_to_factors: a map from time variable to the log prob factors.
+    :param dict time_to_init_vars: a map from time variable to init discrete sites.
+    :param dict time_to_markov_dims: a map from time variable to dimensions at markov sites
+        (discrete sites that depend on previous steps).
+    :param frozenset sum_vars: all plate and enum dimensions in the trace.
+    :param frozenset prod_vars: all plate dimensions in the trace.
+    :returns: a list of factors after eliminate time dimensions
+    """
+    markov_factors = []
+    for time_var, log_factors in time_to_factors.items():
+        prev_vars = time_to_init_vars[time_var]
+        # remove `_init/` prefix to convert prev to curr
+        prev_to_curr = {k: "/".join(k.split("/")[1:]) for k in prev_vars}
+        # we eliminate all plate dimensions not available at discrete sites.
+        eliminate_vars = (sum_vars | prod_vars) - time_to_markov_dims[time_var]
```

Comment:
Does this `eliminate_vars` include variables only enumerated *within* a single timestep in `scan`? For example, suppose that for whatever reason you had a discrete HMM whose emission distributions were Gaussian mixtures rather than Gaussian - you can eliminate the latent mixture component in the emission distribution in the initial `funsor.sum_product.sum_product` call below, but not the latent state of the markov chain.

Reply:
Does this [test](https://github.com/pyro-ppl/numpyro/pull/682/files#diff-fed0396c70de3697c3e0eb518df45b00R303) cover that usage case? Here we not eliminate only dimensions in the latent state of the markov chain.

Reply:
Yes - if you run that test and set a debugger breakpoint here, what are the contents of `eliminate_vars`?  Does it contain `w`?

Reply:
Yes, it is `frozenset({'y', '_init/w', '_init/y', 'D', 'w'}`. Is it expected?

Reply:
Yeah, `w`, `_init/w` and `D` are expected.  `y` shouldn't be in `sum_vars` in the first place, since it's not an input of any of the log-factor funsors. I left a comment indicating where that should be fixed.